Security udpates

Kluny opened the pull request.

Kluny committed:
Escaping and localizing

- Added strict typing to in_array check.
- Added esc_html__() where __() was used.
- Added more checking to $_GET usage.

Kluny committed:
Sanitization and escaping.

Kluny committed:
Sanitization and escaping.

Kluny committed:
Escaping, sanitization, and adding todos.

Kluny committed:
Refactored print_button and print_disconnect_button

- Ensured that redirects are safe.

Kluny committed:
Performance fixes.

- Added wrappers for user_meta methods to allow use of user_attributes when it's availabel on VIP.

Kluny committed:
Escaping.

Kluny committed:
Removed user_met_wrappers; not needed for Go.

Kluny committed:
Removed require_ for user_meta_wrappers.

Kluny commented:
FILE: ...ent/plugins/facebook-login/trunk/includes/class-fbl-upgrader.php
----------------------------------------------------------------------
FOUND 1 ERROR AND 2 WARNINGS AFFECTING 1 LINE
----------------------------------------------------------------------
[ ] 25 | WARNING | Usage of a direct database call is discouraged.
[ ]  25 | WARNING | Direct database call without caching detected.
    |         | Consider using wp_cache_get() / wp_cache_set() or
    |         | wp_cache_delete().
[ ]  25 | ERROR   | Usage of users/usermeta tables is highly discouraged
    |         | in VIP context, For storing user additional user
    |         | metadata, you should look at User Attributes.
----------------------------------------------------------------------



FILE: ...content/plugins/facebook-login/trunk/public/js/facebook-login.js
----------------------------------------------------------------------
FOUND 6 ERRORS AND 2 WARNINGS AFFECTING 3 LINES
----------------------------------------------------------------------
[ ]  53 | WARNING | Any HTML passed to `append` gets executed. Make sure
    |         | it's properly escaped.
[ ]  53 | ERROR   | HTML string concatenation detected: '<p
    |         | class="fbl_error">'+
[ ]  53 | ERROR   | HTML string concatenation detected +'</p>'
[ ]  56 | ERROR   | HTML string concatenation detected: '<p
    |         | class="fbl_error">'+
[ ]  56 | ERROR   | HTML string concatenation detected +'</p>'
[ ]  72 | WARNING | Any HTML passed to `append` gets executed. Make sure
    |         | it's properly escaped.
[ ]  72 | ERROR   | HTML string concatenation detected: '<p
    |         | class="fbl_error">'+
[ ]  72 | ERROR   | HTML string concatenation detected +'</p>'
----------------------------------------------------------------------
TODO: Check all this some more.

FILE: ...gins/facebook-login/trunk/public/class-facebook-login-public.php
----------------------------------------------------------------------
FOUND 22 ERRORS AND 29 WARNINGS AFFECTING 30 LINES
----------------------------------------------------------------------
[ ]  118 | WARNING | Detected access of super global var $_SERVER,
     |         | probably needs manual inspection.
[ ]  118 | ERROR   | Detected usage of a non-validated input variable:
     |         | $_SERVER
[ ]  118 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_SERVER
[ ]  120 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  120 | WARNING | Processing form data without nonce verification.
[ ]  121 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  121 | WARNING | Processing form data without nonce verification.
[ ]  121 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_GET
[ ]  123 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  123 | WARNING | Processing form data without nonce verification.
[ ]  123 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_GET
[ ]  129 | WARNING | Not using strict comparison for in_array; supply
     |         | true for third argument.
[ ]  130 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  130 | WARNING | Processing form data without nonce verification.
[ ]  134 | ERROR   | All output should be run through an escaping
     |         | function (see the Security sections in the WordPress
     |         | Developer Handbooks), found 'apply_filters'.
[ ]  164 | WARNING | Detected access of super global var $_SERVER,
     |         | probably needs manual inspection.
[ ]  164 | ERROR   | Detected usage of a non-validated input variable:
     |         | $_SERVER
[ ]  164 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_SERVER
[ ]  166 | ERROR   | All output should be run through an escaping
     |         | function (see the Security sections in the WordPress
     |         | Developer Handbooks), found 'apply_filters'.
[ ]  167 | WARNING | urlencode() should only be used when dealing with
     |         | legacy applications rawurlencode() should now be
     |         | used instead. See
     |         | http://php.net/manual/en/function.rawurlencode.php
     |         | and http://www.faqs.org/rfcs/rfc3986.html
[ ]  200 | ERROR   | All output should be run through an escaping
     |         | function (see the Security sections in the WordPress
     |         | Developer Handbooks), found 'trim'.
[ ]  230 | ERROR   | All output should be run through an escaping
     |         | function (see the Security sections in the WordPress
     |         | Developer Handbooks), found '$this'.
[ ]  255 | WARNING | Detected access of super global var $_POST, probably
     |         | needs manual inspection.
[ ]  255 | WARNING | Detected access of super global var $_POST, probably
     |         | needs manual inspection.
[ ]  255 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_POST
[ ]  256 | WARNING | Detected access of super global var $_POST, probably
     |         | needs manual inspection.
[ ]  256 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_POST
[ ]  279 | ERROR   | Detected high remote request timeout. `timeout` is
     |         | set to `30`.
[ ]  281 | WARNING | wp_remote_get() is highly discouraged, please use
     |         | vip_safe_wp_remote_get() instead.
[ ]  281 | WARNING | wp_remote_get() is highly discouraged, please use
     |         | vip_safe_wp_remote_get() instead.
[ ]  281 | ERROR   | Detected high remote request timeout. `timeout` is
     |         | set to `30`.
[ ]  409 | WARNING | Variable assignment found within a condition. Did
     |         | you mean to do a comparison?
[ ]  517 | WARNING | Found: !=. Use strict comparisons (=== or !==).
[ ]  531 | WARNING | Found: !=. Use strict comparisons (=== or !==).
[ ]  565 | WARNING | Detected usage of meta_key, possible slow query.
[ ]  566 | WARNING | Detected usage of meta_value, possible slow query.
[ ]  601 | WARNING | Not using strict comparison for in_array; supply
     |         | true for third argument.
[ ]  607 | WARNING | Usage of a direct database call is discouraged.
[ ]  607 | WARNING | Direct database call without caching detected.
     |         | Consider using wp_cache_get() / wp_cache_set() or
     |         | wp_cache_delete().
[ ]  608 | ERROR   | Usage of users/usermeta tables is highly discouraged
     |         | in VIP context, For storing user additional user
     |         | metadata, you should look at User Attributes.
[ ]  662 | ERROR   | All output should be run through an escaping
     |         | function (like esc_html_e() or esc_attr_e()), found
     |         | '_e'.
[ ]  665 | ERROR   | All output should be run through an escaping
     |         | function (like esc_html_e() or esc_attr_e()), found
     |         | '_e'.
[ ]  668 | ERROR   | All output should be run through an escaping
     |         | function (like esc_html_e() or esc_attr_e()), found
     |         | '_e'.
[ ]  685 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  685 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  685 | ERROR   | Detected usage of a non-validated input variable:
     |         | $_GET
[ ]  685 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_GET
[ ]  690 | WARNING | Detected access of super global var $_GET, probably
     |         | needs manual inspection.
[ ]  690 | ERROR   | Detected usage of a non-validated input variable:
     |         | $_GET
[ ]  690 | ERROR   | Detected usage of a non-sanitized input variable:
     |         | $_GET
[ ]  746 | WARNING | Unused variable `$exists`.
----------------------------------------------------------------------

Kluny commented:
FILE: ...ent/plugins/facebook-login/trunk/includes/class-fbl-upgrader.php

FOUND 1 ERROR AND 2 WARNINGS AFFECTING 1 LINE

- [x] 25 | WARNING | Usage of a direct database call is discouraged.
- [x]  25 | WARNING | Direct database call without caching detected. Consider using wp_cache_get() / wp_cache_set() or wp_cache_delete().
- [x]  25 | ERROR   | Usage of users/usermeta tables is highly discouraged in VIP context, For storing user additional user metadata, you should look at User Attributes.
    - checked and it's okay because we're on VIP Go, where user_meta can be used. The database call will only run on upgrade so there's no danger of accidentally impairing site performance. 

FILE: ...content/plugins/facebook-login/trunk/public/js/facebook-login.js

FOUND 6 ERRORS AND 2 WARNINGS AFFECTING 3 LINES

- [x]  53 | WARNING | Any HTML passed to `append` gets executed. Make sure it`s properly escaped.
- [x]  53 | ERROR   | HTML string concatenation detected: `<p class="fbl_error">`+
- [x]  53 | ERROR   | HTML string concatenation detected +`</p>`
- [x]  56 | ERROR   | HTML string concatenation detected: `<p class="fbl_error">`+
- [x]  56 | ERROR   | HTML string concatenation detected +`</p>`
- [x]  72 | WARNING | Any HTML passed to `append` gets executed. Make sure it`s properly escaped.
- [x]  72 | ERROR   | HTML string concatenation detected: `<p class="fbl_error">`+
- [x]  72 | ERROR   | HTML string concatenation detected +`</p>`

TODO: Check all this some more.

FILE: ...gins/facebook-login/trunk/public/class-facebook-login-public.php

FOUND 22 ERRORS AND 29 WARNINGS AFFECTING 30 LINES

- [x]  118 | WARNING | Detected access of super global var $_SERVER, probably needs manual inspection.
- [x]  118 | ERROR   | Detected usage of a non-validated input variable: $_SERVER
- [x]  118 | ERROR   | Detected usage of a non-sanitized input variable: $_SERVER
- [x]  120 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  120 | WARNING | Processing form data without nonce verification.
- [x]  121 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  121 | WARNING | Processing form data without nonce verification.
- [x]  121 | ERROR   | Detected usage of a non-sanitized input variable: $_GET
- [x]  123 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  123 | WARNING | Processing form data without nonce verification.
- [x]  123 | ERROR   | Detected usage of a non-sanitized input variable: $_GET
- [x]  129 | WARNING | Not using strict comparison for in_array; supply true for third argument.
- [x]  130 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  130 | WARNING | Processing form data without nonce verification.
- [x]  134 | ERROR   | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found `apply_filters`.
- [x]  164 | WARNING | Detected access of super global var $_SERVER, probably needs manual inspection.
- [x]  164 | ERROR   | Detected usage of a non-validated input variable: $_SERVER
- [x]  164 | ERROR   | Detected usage of a non-sanitized input variable: $_SERVER
- [x]  166 | ERROR   | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found `apply_filters`.
- [x]  167 | WARNING | urlencode() should only be used when dealing with legacy applications rawurlencode() should now be used instead. See http://php.net/manual/en/function.rawurlencode.php and http://www.faqs.org/rfcs/rfc3986.html
- [x]  200 | ERROR   | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found `trim`.
- [x]  230 | ERROR   | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found `$this`.
- [x]  255 | WARNING | Detected access of super global var $_POST, probably needs manual inspection.
- [x]  255 | WARNING | Detected access of super global var $_POST, probably needs manual inspection.
- [x]  255 | ERROR   | Detected usage of a non-sanitized input variable: $_POST
- [x]  256 | WARNING | Detected access of super global var $_POST, probably needs manual inspection.
- [x]  256 | ERROR   | Detected usage of a non-sanitized input variable: $_POST
- [x]  279 | ERROR   | Detected high remote request timeout. `timeout` is set to `30`.
- [x]  281 | WARNING | wp_remote_get() is highly discouraged, please use vip_safe_wp_remote_get() instead.
- [x]  281 | WARNING | wp_remote_get() is highly discouraged, please use vip_safe_wp_remote_get() instead.
- [x]  281 | ERROR   | Detected high remote request timeout. `timeout` is set to `30`.
- [x]  409 | WARNING | Variable assignment found within a condition. Did you mean to do a comparison?
- [x]  517 | WARNING | Found: !=. Use strict comparisons (=== or !==).
- [x]  531 | WARNING | Found: !=. Use strict comparisons (=== or !==).
- [x]  565 | WARNING | Detected usage of meta_key, possible slow query.
- [x]  566 | WARNING | Detected usage of meta_value, possible slow query.
- [x]  601 | WARNING | Not using strict comparison for in_array; supply true for third argument.
- [x]  607 | WARNING | Usage of a direct database call is discouraged.
- [x]  607 | WARNING | Direct database call without caching detected. Consider using wp_cache_get() / wp_cache_set() or wp_cache_delete().
- [x]  608 | ERROR   | Usage of users/usermeta tables is highly discouraged in VIP context, For storing user additional user metadata, you should look at User Attributes.
- [x]  662 | ERROR   | All output should be run through an escaping function (like esc_html_e() or esc_attr_e()), found `_e`.
- [x]  665 | ERROR   | All output should be run through an escaping function (like esc_html_e() or esc_attr_e()), found `_e`.
- [x]  668 | ERROR   | All output should be run through an escaping function (like esc_html_e() or esc_attr_e()), found `_e`.
- [x]  685 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  685 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  685 | ERROR   | Detected usage of a non-validated input variable: $_GET
- [x]  685 | ERROR   | Detected usage of a non-sanitized input variable: $_GET
- [x]  690 | WARNING | Detected access of super global var $_GET, probably needs manual inspection.
- [x]  690 | ERROR   | Detected usage of a non-validated input variable: $_GET
- [x]  690 | ERROR   | Detected usage of a non-sanitized input variable: $_GET
- [x]  746 | WARNING | Unused variable `$exists`.

Kluny committed:
Added escaping, checked redirect urls for safety, added strict typing.

Added comments explaining database calls.

nicholasio commented on trunk/admin/class-facebook-login-admin.php:
```diff
@@ -112,13 +114,13 @@ public function admin_scripts() {
 	 */
 	public function profile_buttons( $user ) {
-		?><h3><?php _e("Facebook connection", "blank"); ?></h3><?php
+		?><h3><?php esc_html_e("Facebook connection", "blank"); ?></h3><?php
```
Why is the text domain "blank"? I think it should be 'fbl'

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
 		// if we are in login page we don't want to redirect back to it
-		if ( isset( $GLOBALS['pagenow'] ) && in_array( $GLOBALS['pagenow'], array( 'wp-login.php', 'wp-register.php' ) ) && empty($_GET['redirect_to']) )
+		if ( isset( $GLOBALS['pagenow'] )
+             && in_array( $GLOBALS['pagenow'], array( 'wp-login.php', 'wp-register.php' ), true )
```
Looks like your indentation is off here.

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
+
+		echo apply_filters('fbl/login_button',
+            '<div class="fbl-button" 
+                        data-redirect="' . esc_attr( apply_filters( 'flp/redirect_url', $redirect ) ) . '" 
```
I think you'll need to pass $redirect through `rawurlencode`

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
-		echo apply_filters('fbl/disconnect_button', '<a href="?fbl_disconnect&fb_nonce='. wp_create_nonce( 'fbl_disconnect' ) .'&redirect='.urlencode( $redirect ).'" class="css-fbl "><div>'. __('Disconnect Facebook', 'fbl') .'<img data-no-lazy="1" src="'.site_url('/wp-includes/js/mediaelement/loading.gif').'" alt="" style="display:none"/></div></a>');
+		echo apply_filters('fbl/disconnect_button',
+            '<a href="?fbl_disconnect&fb_nonce='. wp_create_nonce( 'fbl_disconnect' ) .'&redirect='. rawurlencode( $redirect ).'" class="css-fbl ">
+                <div>'. __('Disconnect Facebook', 'fbl') .'
```
use esc_html__ here

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -159,7 +198,7 @@ public function add_fb_scripts(){
 			    try{
 	                window.FB.init({
-	                    appId      : '<?php echo trim( $this->opts['fb_id'] );?>',
+	                    appId      : '<?php echo trim( esc_attr( $this->opts['fb_id'] ) );?>',
```
Use `esc_js` instead.

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -187,7 +226,9 @@ function fbl_init(){
-				js.src = "//connect.facebook.net/<?= $this->locale;?>/sdk.js";
+				<?php   // @vicpcs: Value of $this->locale is thoroughly escaped elsewhere
+                        // in this file; see set_locale(). ?>
+				js.src = "//connect.facebook.net/<?php echo $this->locale; ?>/sdk.js";
```
You can add a comment like this to disable phpcs for this line: // WPCS: XSS Ok.

Then you can explain why it is good.

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -212,8 +253,8 @@ public function jetpack_script_args( $args ){
 
-		$access_token = isset( $_POST['fb_response']['authResponse']['accessToken'] ) ? $_POST['fb_response']['authResponse']['accessToken'] : '';
-		$fb_user_id = $_POST['fb_response']['authResponse']['userID'];
+		$access_token = isset( $_POST['fb_response']['authResponse']['accessToken'] ) ? sanitize_text_field()( $_POST['fb_response']['authResponse']['accessToken'] ) : '';
```
You call to sanitize_text_field is incorrect. `sanitize_text_field()( $_POST['fb_response']['authResponse']['accessToken'] )`

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -295,8 +340,8 @@ public function login_or_register_user() {
 			if( !$meta_updated )
-				update_user_meta( $user_id, '_fb_user_id', $user['fb_user_id'] );
-			do_action( 'fbl/after_login', $user, $user_id);
+				delete_user_meta( $user_id, '_fb_user_id', $user['fb_user_id'] );
```
Why did you change this to delete_user_meta? I think this needs to be update_user_meta as it receives 3 parameters.

Kluny replied:
I started refactoring *_user_meta to *_user_attribute, then realized it's not relevant for Go. Did it with a find-replace, so I'm not sure. Fixed!

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -645,7 +690,16 @@ public function disconnect_facebook( ) {
+            }
+        }
+        
+		wp_redirect( $redirect );
```
use `wp_safe_redirect` instead.

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -645,7 +690,16 @@ public function disconnect_facebook( ) {
 		// refresh page
-		wp_redirect( esc_url( $_GET['redirect'] ) );
+
+        $redirect = home_url();
```
Indentation here is off.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -645,7 +690,16 @@ public function disconnect_facebook( ) {
-		wp_redirect( esc_url( $_GET['redirect'] ) );
+
+        $redirect = home_url();
+        if( ! empty( $_GET['redirect'] ) ) {
```
Can you elaborate on why we have $_GET['redirect'] and $_GET['redirect_to']?

Kluny replied:
`redirect_to` appears to be WordPress convention, `redirect` appears to be only used in the disconnect button form. I'm in favor of changing the plugin as little as possible beyond security and plugin updates,  but it is inconsistent. Changed it. 
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
 	 */
 	public function print_button() {
-		$redirect = ! empty( $_GET['redirect_to'] ) ? esc_url($_GET['redirect_to']) : ( is_ssl() ? 'https://' : 'http://' ) . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
+		$redirect = home_url() . '/' . $_SERVER['REQUEST_URI']; // @nicholas we discussed this, but still not sure this is sufficient/correct. Opinion?
```
You can replace all of the code you added with something like this: 
`$redirect_to = get_home_url( 1, $_SERVER['REQUEST_URI'] );`

We'll check the URL before performing the actual login.

nicholasio replied:
Actually, after reviewing the plugin code. For added security. Let have the ajax callback return the redirect_to url. Because right now the JS code is going to redirect to whatever is set in data-redirect which I don't like (https://github.com/Kluny/facebook-login/blob/master/trunk/public/js/facebook-login.js#L41). It wouldn't be hard for some script to trick the user and redirect it to somewhere else.

So return the actual redirect_to url from as part of the ajax response of this method: https://github.com/Kluny/facebook-login/blob/3f02f8c3f8337adc1b451ed430b71623224283b3/trunk/public/class-facebook-login-public.php#L346

This will make things more secure.

Kluny replied:
I've changed this a bit. I found out that the redirect wasn't working at all. Eg if you're on https://fbelearning.ngrok.io/post-21/, it would redirect you to https://fbelearning.ngrok.io/. I moved this part:

``` $redirect = home_url() . '/' . $_SERVER['REQUEST_URI']; // @nicholas we discussed this, but still not sure this is sufficient/correct. Opinion?
-
-        if( ! empty( $_GET['redirect_to'] ) ) {
-               $server = strpos( $_GET['redirect_to'], home_url() );
-            if( $server === 0 ) {
-                $redirect = esc_url( $_GET['redirect_to'] );
-            }
-        }
```

to register_or_login_user(), but then it redirect users to /wp-admin/ajax.php. I changed it to use HTTP_REFERER, and now it works right. Referrer is still spoofable, so there's still a check for the site_url within it. Tests okay.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -645,7 +690,16 @@ public function disconnect_facebook( ) {
-		wp_redirect( esc_url( $_GET['redirect'] ) );
+
+        $redirect = home_url();
+        if( ! empty( $_GET['redirect'] ) ) {
```
I'd suggest to do something like this to check if the URL is a valid URL:

```
	/**
	 * Checks if we need to redirect user
	 *
	 * @return void
	 */
	protected function maybe_redirect_user() {
		$redirect_to = filter_input( INPUT_GET, 'redirect_to', FILTER_SANITIZE_URL );
		$home_url = home_url();

		if ( empty( $redirect_to ) ) {
			$redirect_to = $home_url;
		}

		$parsed_redirect_to = parse_url( $redirect_to );
		$parsed_home_url    = parse_url( $home_url );
		// only allow redirects to same host
		if ( $parsed_redirect_to['host'] === $parsed_home_url['host'] ) {
			wp_safe_redirect( $redirect_to );
		} else {
			wp_safe_redirect( $home_url );
		}

		exit;
	}
```

Kluny replied:
👍

nicholasio commented on trunk/includes/class-fbl-upgrader.php:
```diff
@@ -20,9 +20,11 @@ public function upgrade_plugin() {
+
+		// to prevent unauthorized access, delete all fb_user_ids
 		if ( ! empty( $current_version ) && version_compare( $current_version, '1.0.7.2', '<' ) &&  version_compare( $current_version, '1.0.5', '>' ) ) {
 			$wpdb->query( "DELETE FROM $wpdb->usermeta WHERE meta_key = '_fb_user_id'");
```
Disable the Code Sniffer for this line, then we'll not have it popup in their automated sniffers.

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
 	 * @since 1.1
 	 */
 	public function print_disconnect_button( ) {
+		$redirect = apply_filters( 'flp/disconnect_redirect_url', home_url() . '/' . $_SERVER['REQUEST_URI'] );
```
You can replace the URL part with `get_home_url( 1, $_SERVER['REQUEST_URI'] );` for consistency.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -115,26 +115,65 @@ public function enqueue_scripts() {
+            '<a href="?fbl_disconnect&fb_nonce='. wp_create_nonce( 'fbl_disconnect' ) .'&redirect='. rawurlencode( $redirect ).'" class="css-fbl ">
+                <div>'. __('Disconnect Facebook', 'fbl') .'
+                    <img data-no-lazy="1" 
+                            src="' . site_url('/wp-includes/js/mediaelement/loading.gif') . '" 
```
pass site_url through esc_url()

Kluny replied:
👍

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -187,7 +226,9 @@ function fbl_init(){
 				js = d.createElement(s); js.id = id;
-				js.src = "//connect.facebook.net/<?= $this->locale;?>/sdk.js";
+				<?php   // @vicpcs: Value of $this->locale is thoroughly escaped elsewhere
+                        // in this file; see set_locale(). ?>
```
indentation a bit off here.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -556,12 +599,14 @@ private function generateUsername( $user ) {
+		// "Generate" unique suffix. Finds out how many existing users are in the database with the given username (minus numerical suffix), and appends that number to the new username as a pseudounique suffix.
+        // Eg. "SELECT 1 + SUBSTR(user_login, 9) FROM wp_users WHERE user_login REGEXP '^shannon(-[0-9]+)?$' ORDER BY 1 DESC LIMIT 1".
+        // If user 'shannon' exists, result is one field in one row containing "1". That number is appended to username.
 		$suffix = $wpdb->get_var( $wpdb->prepare(
```
This is actually not that important. we can just have any unique id, what really matters is the first and last name. The user will never login using his email or username, they'll always login through facebook. So we can replace this complex logic with something more simple

Kluny replied:
Yes, but if we are asking the owner of this project to merge in our changes, they may not agree. Ravi had a suggestion here for changing the logic  - https://10up.slack.com/archives/C105936H3/p1528779474000233 - it's not simpler but it doesn't do that heavy database call. What you think?

nicholasio replied:
Right, let's add Ravi's solution for now.

Kluny replied:
Done. 👍

Kluny committed:
Changed redirect for better security

Redirect URL is checked to ensure that it's the same host as the site within php to prevent malicious js redirection.

Other changes as suggested in code review
	- fixed incorrect method call
	- changed _() to esc_html__()
	- fixed textdomain
	- fixed indentation
	- turned off codesniffer for certain lines
	- added comments for wpcs

Kluny committed:
Chagned diconnect form to use redirect_to consistent with WP

... instead of `redirect`.

Kluny committed:
Changed js to use DOM api to append error message.

Kluny committed:
Changed database-heavy call for generating username

- Used @ravichandrach1's method to increment username suffix.
- Tested using 2-3 facebook dummy accounts.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -645,7 +713,17 @@ public function disconnect_facebook( ) {
+
+		$redirect = home_url();
+		$parsed_home_url = parse_url( home_url() );
+		if( ! empty( $_GET['redirect_to'] ) ) {
```
Move this code to a separate method, it is being used multiple time.

nicholasio replied:
and receive redirect_to as a parameter.

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -212,8 +237,22 @@ public function jetpack_script_args( $args ){
+			if( $parsed_get_url['host'] === $parsed_home_url['host'] ) {
+				$redirect = esc_url( $_GET['redirect_to'] );
+			}
+		} else if( ! empty( $_SERVER['HTTP_REFERER'] ) ) {
```
HTTP_REFERER is not safe, I know we're checking it but is there any reason to check for it? IN case a user is logging from a page other than the homepage, passing redirect_to should redirect appropriately.

Kluny replied:
In this block, the elseif is getting executed, not the if.

```
        $redirect = home_url();
        $parsed_home_url = parse_url( home_url() );
        if( ! empty( $_GET['redirect_to'] ) ) {
			$parsed_get_url = parse_url( $_GET['redirect_to'] );
			if( $parsed_get_url['host'] === $parsed_home_url['host'] ) {
				$redirect = esc_url( $_GET['redirect_to'] );
			}
		} else if( ! empty( $_SERVER['HTTP_REFERER'] ) ) {
			$parsed_referer = parse_url( $_SERVER['HTTP_REFERER'] );
			if( $parsed_referer['host'] === $parsed_home_url['host'] ) {
				$redirect = esc_url( $_SERVER['HTTP_REFERER'] );
			}
		}
```

 So it seems like http_referer is needed as it's written - redirect_to is passed if a user tries to access a page that's only available to logged in users, such as wp-admin. It would be better to only use $redirect_to, though. Maybe that's why $_GET['redirect'] was being used before, to pass it to Facebook. I'll see if I can get it to pass redirect_to instead.

Kluny replied:
I passed the referer to js as $wp->request. That ought to be safe.

nicholasio reviewed (changes requested):
Additional Feedback:

- This line in assets/js can be removed: https://github.com/Kluny/facebook-login/blob/48c3530c1e721a099aeec49e5c7e2fce9c23611b/trunk/public/js/facebook-login.js#L16
- Update the plugin version in README;[txt|md] and add a line to changelog summarizing the changes (e.g various security fixes to meet VIP rules)

Kluny committed:
Removed unneeded variable declaration.

Kluny committed:
Passed current page url to js through l8n script...

To ensure safe and accurate redirection.

Kluny committed:
Moved redirect code to a method.

Kluny committed:
Tested and refactored redirect methd to use wp_get_referer to accurately redirect users.

Kluny committed:
Updated version number and changelog.

Kluny commented:
I've made those changes, Nicholas!

- removed that js variable
- moved the redirect code to a method
- changed it to it's not using http_referrer anymore - the GET wasn't working the way it was supposed to, changed that to wp_get_referrer and now it works as expected
- updated version and changelog

nicholasio commented on trunk/public/class-facebook-login-public.php:
```diff
@@ -206,14 +233,37 @@ public function jetpack_script_args( $args ){
+		$parsed_home_url = parse_url( home_url() );
+		
+		$referrer = parse_url( urldecode( wp_get_referer() ) );
+        if(! empty( $referrer['query']) ) {
```
Spacing is off again.

nicholasio commented on trunk/public/js/facebook-login.js:
```diff
@@ -38,22 +37,28 @@ function fbl_loginCheck() {
                 success: function (data) {
                     if (data && data.success) {
-                        if( data.redirect && data.redirect.length ) {
+                        if ( fbl.current_page && fbl.current_page.length ) {
```
We're close! 

You could pass `fbl.current_page` to the ajax call and then have it validate the redirect there, this would make things simpler and actually leverage `redirect_to`, I'm little unsure if redirect_to would ever be used as it is right now

Kluny committed:
Whitespace.
